### DISCUSSION TITLE
`useEffect` cleanup function is being called on mount and not on unmount

### DISCUSSION THREAD

**Discussion by hjonasson** [vercel/next.js#50515] [Help] [answered]
### Summary

I have a database I want to clean up when a user leaves the page. I thought I would do that with a `useEffect` cleanup function like so:

```TS
// app/page.tsx

export default function MyComponent() {
  useEffect(() => {
    const cleanup = async () => {
      await fetch("");
    };
    return () => {
      cleanup();
    };
  }, []);
  return <div>Some component</div>;
}
```

It turns out that the cleanup function is called on mount, which I didn't expect and not on unmount, where I did expect it. Does anyone have any pointers about how to debug this?

Thanks!

PS: I am open to any other way to make an API call on unmount/user leaving. `router.beforepopstate` existed in the `pages` directory but not in the `app` directory. I did not find a similar method there.

### Additional information

```js
// Put this into any NextJS project's `app/page.tsx` file and see the console log running on page load.

// app/page.tsx

export default function MyComponent() {
  useEffect(() => {
    return () => {
      console.log("Firing on mount and not on cleanup");
    };
  }, []);
  return <div>Some component</div>;
}
```

**Comment by icyJoseph** [comment, marked as answer]
Hi,

Are you perhaps being tricked by strict mode? Could you confirm if it is activated? It is in the `next.config.js` file, called `reactStrictMode`.

Strict Mode, in dev mode, forces an additional mount/unmount cycle, which might explain why you think it is working on mount. Rather, the page is mounted->unmounted->"mounted again", and you are seeing the "mounted again" step. 

For what is worth, I tried this and I did not see the same behaviour.

- https://legacy.reactjs.org/docs/strict-mode.html#ensuring-reusable-state
- https://react.dev/reference/react/StrictMode#fixing-bugs-found-by-double-rendering-in-development

**Comment by hjonasson** [comment]
Thanks for looking at this issue @icyJoseph! Strict mode is indeed on. Thanks for saving us a lot of time.
